### PR TITLE
Duplicate service port name can't be validated

### PR DESCRIPTION
**Pull request by feihujiang**
#7485

### CODE REVIEW

**Comment by thockin** [comment]
Nice catch!  can you add a test case for this, please?

**Comment by feihujiang** [comment]
@thockin  I found there is a test case named "dup port name" in the TestValidateService function , but there are three small bugs,  and I fixed them. I think it's not necessary to add a test case, do you think so?

File: pkg/api/validation/validation_test.go [hunk 1]
```diff
@@ -1674,7 +1674,7 @@ func TestValidateService(t *testing.T) {
 			name: "dup port name",
 			tweakSvc: func(s *api.Service) {
 				s.Spec.Ports[0].Name = "p"
-				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "p", Port: 12345})
+				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "p", Port: 12345, Protocol: "TCP"})
```
Comment (thockin) [inline pkg/api/validation/validation_test.go]: good catch

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -1690,7 +1690,7 @@ func TestValidateService(t *testing.T) {
 			name: "invalid load balancer protocol 2",
 			tweakSvc: func(s *api.Service) {
 				s.Spec.CreateExternalLoadBalancer = true
-				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "p", Port: 12345, Protocol: "UDP"})
+				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "ilbp2", Port: 12345, Protocol: "UDP"})
```
Comment (thockin) [inline pkg/api/validation/validation_test.go]: this change seems unrelated - can you revert this piece

File: pkg/api/validation/validation_test.go [hunk 3]
```diff
@@ -1742,7 +1742,7 @@ func TestValidateService(t *testing.T) {
 			name: "valid external load balancer 2 ports",
 			tweakSvc: func(s *api.Service) {
 				s.Spec.CreateExternalLoadBalancer = true
-				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "p", Port: 12345, Protocol: "TCP"})
+				s.Spec.Ports = append(s.Spec.Ports, api.ServicePort{Name: "velb2p", Port: 12345, Protocol: "TCP"})
```
Comment (thockin) [inline pkg/api/validation/validation_test.go]: and this

File: pkg/api/validation/validation_test.go (hunk 2 above)
Comment (feihujiang) [inline pkg/api/validation/validation_test.go]: The s.spec.Ports has two ServicePort instances, one is defined in the makeValidService() function, the name of the first one is also "p", so I changed the name of the second one.

File: pkg/api/validation/validation_test.go (hunk 3 above)
Comment (feihujiang) [inline pkg/api/validation/validation_test.go]: The s.spec.Ports have two ServicePort instances, one is defined in the makeValidService() function, the name of the first one is also "p", so I changed the name of the second one.

**Comment by feihujiang** [comment]
@thockin I add some notes. Could you check them, please?

**Comment by feihujiang** [comment]
I think thockin may be busy. @cjcullen Could you check them please? Thanks!

File: pkg/api/validation/validation_test.go (hunk 2 above)
Comment (thockin) [inline pkg/api/validation/validation_test.go]: instead of "ilbp2" which seems random (I know it is the test case), can this just be "q" - we use "p" and "q" for a lot of two-port cases.

File: pkg/api/validation/validation_test.go (hunk 3 above)
Comment (thockin) [inline pkg/api/validation/validation_test.go]: same comment - "q" ?

**Comment by thockin** [comment]
Sorry, was out for almost a week.  LGTM but for the two name nits.

File: pkg/api/validation/validation_test.go (hunk 2 above)
Comment (feihujiang) [inline pkg/api/validation/validation_test.go]: Ok, I changed it

File: pkg/api/validation/validation_test.go (hunk 3 above)
Comment (feihujiang) [inline pkg/api/validation/validation_test.go]: I changed it too.

**Comment by feihujiang** [comment]
@thockin Ok, I changed the two names. Could you please review them again?

**Comment by thockin** [comment]
Thanks!

**Comment by nikhiljindal** [comment]
Needs a rebase for merging.
@feihujiang Can you please rebase?

**Comment by feihujiang** [comment]
@nikhiljindal  Yes, I rebased.
